src/brod_producer.erl

```diff
@@ -145,61 +160,68 @@ sync_produce_request(#brod_produce_reply{call_ref = CallRef} = Reply) ->
 %%%_* gen_server callbacks =====================================================
 
 init({ClientPid, Topic, Partition, Config}) ->
-  self() ! init_socket,
-  {ok, #state{ client_pid = ClientPid
-             , topic      = Topic
-             , partition  = Partition
-             , config     = Config
-             }}.
-
-handle_info(init_socket, #state{ client_pid = ClientPid
-                               , topic      = Topic
-                               , partition  = Partition
-                               , config     = Config0
-                               } = State) ->
-  %% 1. Lookup, or maybe (re-)establish a connection to partition leader
-  {ok, SockPid} =
-    brod_client:get_leader_connection(ClientPid, Topic, Partition),
-  _ = erlang:monitor(process, SockPid),
+  BufferLimit = ?config(partition_buffer_limit, ?DEFAULT_PARITION_BUFFER_LIMIT),
+  OnWireLimit = ?config(partition_onwire_limit, ?DEFAULT_PARITION_ONWIRE_LIMIT),
+  MaxBatchSize = ?config(max_batch_size, ?DEFAULT_MAX_BATCH_SIZE),
+  RequiredAcks = ?config(required_acks, ?DEFAULT_REQUIRED_ACKS),
+  AckTimeout = ?config(ack_timeout, ?DEFAULT_ACK_TIMEOUT),
 
-  %% 2. Register self() to client.
-  ok = brod_client:register_producer(ClientPid, Topic, Partition),
-
-  %% 3. Create the producing buffer
-  {BufferLimit, Config1} = take_config(partition_buffer_limit,
-                                       ?DEFAULT_PARITION_BUFFER_LIMIT,
-                                       Config0),
-  {OnWireLimit, Config2} = take_config(partition_onwire_limit,
-                                       ?DEFAULT_PARITION_ONWIRE_LIMIT,
-                                       Config1),
-  {MsgSetBytes, Config}  = take_config(message_set_bytes_limit,
-                                       ?DEFAULT_MESSAGE_SET_BYTES_LIMIT,
-                                       Config2),
-  RequiredAcks = get_required_acks(Config),
-  AckTimeout = get_ack_timeout(Config),
   SendFun =
-    fun(KafkaKvList) ->
-      Data = [{Topic, [{Partition, KafkaKvList}]}],
-      KafkaReq = #produce_request{ acks    = RequiredAcks
-                                 , timeout = AckTimeout
-                                 , data    = Data
-                                 },
-      case brod_sock:send(SockPid, KafkaReq) of
-        ok              -> ok;
-        {ok, CorrId}    -> {ok, CorrId};
-        {error, Reason} -> exit(Reason)
-      end
+    fun(SockPid, KafkaKvList) ->
+        Data = [{Topic, [{Partition, KafkaKvList}]}],
+        KafkaReq = #produce_request{ acks    = RequiredAcks
+                                   , timeout = AckTimeout
+                                   , data    = Data
+                                   },
+        case brod_sock:send(SockPid, KafkaReq) of
+          ok              -> ok;
+          {ok, CorrId}    -> {ok, CorrId};
+          {error, Reason} -> {error, Reason}
```

Comment:
Reason is always {sock_down, SockDownReason}.

 i don't see how it's possible to retry now neither in the future. why don't keep it simple ?